### ISSUE TITLE
systemd-nspawn+macvlan "could not bring up interface: Device or resource busy"

### ISSUE THREAD

**Issue by SjonHortensius** [systemd/systemd#680]
I use systemd-nspawn with macvlan on a few different machines; and all of them suffer the same issue. Containers (which boot successfully) cannot be restarted; their network-connection will fail 99% of the time. I use the following overload:

```
# cat /etc/systemd/system/systemd-nspawn\@.service.d/start-network-macvlan.conf 
[Service]
ExecStart=
ExecStart=/usr/bin/systemd-nspawn --quiet --keep-unit --boot --link-journal=guest --network-macvlan=br-lan --machine=%I
```

Sometimes the reboot works (at least with v221); but most of the times it errors with either:

```
systemd-networkd[29]: mv-br-lan: Cannot configure IPv4 forwarding for interface mv-br-lan: Read-only file system
systemd-networkd[29]: Enumeration completed
systemd-networkd[29]: mv-br-lan: mv-br-lan       : could not bring up interface: Device or resource busy
systemd-networkd[29]: mv-br-lan: mv-br-lan       : could not set route: Network is unreachable
systemd-networkd[29]: mv-br-lan: Configured
```

or (another machine):

```
systemd-networkd[23]: mv-enp0s25: Cannot configure IPv4 forwarding for interface mv-enp0s25: Read-only file system
systemd-networkd[23]: mv-enp0s25: Cannot configure IPv6 forwarding for interface: Read-only file system
systemd-networkd[23]: Enumeration completed
systemd-networkd[23]: mv-enp0s25: mv-enp0s25      : could not bring up interface: Device or resource busy
```

This still happens with systemd-222, is there any way for me to debug this? Right now I have to restart the whole host to get the container to function again. I think something related to cgroups or the vlan hangs after the container is stopped; preventing a successful restart

**Comment by poettering** [comment]
Hmm, so this only happens during boot?

There's a race between nspawn and networkd: nspawn with the command line above will expect the bridge to be up, but networkd might still setting it up. We can fix that eventually by making nspawn talk to networkd via the bus, and syncing on creating of the bridge.

**Comment by SjonHortensius** [comment]
No it actually only works on first boot! The containers cannot be restarted after that first boot; they will never have a working network connection.

Also, the other machine does macvlan without a bridge and has the same issue

**Comment by poettering** [comment]
what do you mean by "after first boot"? "on subsequent boots"? or "during normal runtime"?

**Comment by SjonHortensius** [comment]
When the host boots, all containers start up fine. If I execute `systemctl reboot` in the container, or `machinectl restart ...` on the host or `systemctl restart systemd-nspawn@...` (or stop; start) on the host; the container won't boot.

So "on subsequent boots" of the container but during normal runtime of the host.

**Comment by SjonHortensius** [comment]
This seems to be fixed in 224, were there any changes done that might have caused that? I'll be testing to make sure it really is fixed

**Comment by arjenschol** [comment]
I've tested this issue and it looks like this is still a problem.
And I think this thread discusses the same issue: http://lists.freedesktop.org/archives/systemd-devel/2015-May/031427.html

**Comment by f0** [comment]
we see the same error, running coreos (systemd v225) and centos 7.2  (systemd 219) on container site with systemd-networkd

this is the log from inside the container after to get this output i must restart systemd-networkd

```
Jan 20 21:59:13 memcache03.1example.net systemd[1]: Starting Network Service...
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interfac
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Got message type=method_return sender=org.freedesktop.DBus destination=:1.7 object=n/a interface=n/a member=n/a
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interfac
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=method_call sender=n/a destination=org.freedesktop.DBus object=/org/freedesktop/DBus interfac
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: timestamp of '/etc/systemd/network' changed
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: timestamp of '/usr/lib/systemd/network' changed
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : flags change: +MULTICAST +BROADCAST
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : link 2 added
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : link state is up-to-date
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : found matching network '/etc/systemd/network/eno3-2192.network'
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : Cannot configure IPv4 forwarding for interface mv-eno3-2192: Read-only file system
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : Cannot configure IPv6 forwarding for interface: Read-only file system
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.fre
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : bringing link up
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.fre
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : setting addresses
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : saved original MTU: 1500
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.fre
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : link 1 added
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : link state is up-to-date
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : unmanaged
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_31 interface=org.fre
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : saved original MTU: 0
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : Adding address: ::1/128 (valid for ever)
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : Adding address: 10.62.192.26/26 (valid for ever)
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: lo              : Adding address: 127.0.0.1/8 (valid for ever)
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Enumeration completed
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : mv-eno3-2192    : could not bring up interface: Device or resource busy
Jan 20 21:59:13 memcache03.1example.net systemd[1]: Started Network Service.
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : Updating address: 10.62.192.26/26 (valid for ever)
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : addresses set
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: Sent message type=signal sender=n/a destination=n/a object=/org/freedesktop/network1/link/_32 interface=org.fre
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : setting routes
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : mv-eno3-2192    : could not set route: Network is unreachable
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : routes set
Jan 20 21:59:13 memcache03.1example.net systemd-networkd[215]: mv-eno3-2192    : link configured


```

**Comment by joejulian** [comment]
I'm seeing the same issue with 231 (Arch) host and a 219 (CentOS 7) guest. First time the machine is started, the link works fine. Power off the machine and start the machine (`machinectl poweroff test1; machinectl start test1`) and the link will not come up.

```
# ip -d link show mv-br0
2: mv-br0: <BROADCAST,MULTICAST> mtu 9000 qdisc noop state DOWN mode DEFAULT qlen 1000
    link/ether aa:13:9f:1c:7a:ef brd ff:ff:ff:ff:ff:ff link-netnsid 0 promiscuity 0 
    macvlan  mode bridge addrgenmode eui64
```

Restarting systemd-networkd has no affect (either the host nor the guest).

Removing the host's macvlan bridge (br0 in this example) and restarting systemd-networkd to recreate it does allow the guest's link to function again until the guest is restarted.

**Comment by gtjoseph** [comment]
Same issue with 232 on Fedora 25/kernel 4.8.10.  Using macvlan, the first time the container is started after a host reboot the network interface comes up.  If the container is shut down and restarted, the network interface does NOT come back up and attempts to bring it up fail.

```
# ifconfig -a
lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1  (Local Loopback)
        RX packets 7  bytes 2345 (2.2 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 7  bytes 2345 (2.2 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

mv-eno1.172: flags=4098<BROADCAST,MULTICAST>  mtu 1500
        inet 172.31.1.22  netmask 255.255.0.0  broadcast 172.31.255.255
        ether 0a:40:d8:77:bc:49  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

# ifconfig mv-eno1.172 up
SIOCSIFFLAGS: Device or resource busy
```
Note that systemd-networkd actually configured the interface, but just can't bring it up.  Oddly enough, container will respond to ARP and ICMP requests but since the interface is down, none of the container's daemons will bind to it.

HOWEVER, it's not really a host reboot that's needed to fix it.  It's just rmmoding macvlan on the host and restarting the container that allows the container to bring up the interface.

**Comment by SjonHortensius** [comment]
With 232 the error is slightly different (I've also added `--network-veth` to the nspawn, not sure if that causes this); attempting to execute `reboot` from container, the host will log:

```
systemd-nspawn[661]: Rebooting.
kernel: IPVS: Creating netns size=2176 id=7
systemd-nspawn[661]: Failed to add new veth interfaces (vb-xxx:host0): File exists
systemd-networkd[612]: vb-xxx: Lost carrier
systemd-timesyncd[585]: Network configuration changed, trying to establish connection.
kernel: DEV: port 6(vb-xxx) entered disabled state
kernel: device vb-xxx left promiscuous mode
kernel: DEV: port 6(vb-xxx) entered disabled state
```

after `systemctl restart systemd-nspawn@xxx` on host:

```
systemd[1]: Stopped Container xxx.
systemd[1]: Starting Container xxx...
kernel: IPVS: Creating netns size=2176 id=8
systemd-timesyncd[585]: Network configuration changed, trying to establish connection.
kernel: IPv6: ADDRCONF(NETDEV_UP): vb-xxx: link is not ready
kernel: DEV: port 6(vb-xxx) entered blocking state
kernel: DEV: port 6(vb-xxx) entered disabled state
kernel: device vb-xxx entered promiscuous mode
systemd-nspawn[26546]: Failed to register machine: Machine 'xxx' already exists
```

after `systemctl stop systemd-nspawn@xxx; systemctl start systemd-nspawn@xxx` on host:
```
# stop
systemd[1]: Stopped Container xxx.
# start
systemd[1]: Starting Container xxx...
kernel: IPVS: Creating netns size=2176 id=9
systemd-timesyncd[585]: Network configuration changed, trying to establish connection.
kernel: IPv6: ADDRCONF(NETDEV_UP): vb-xxx: link is not ready
kernel: DEV: port 6(vb-xxx) entered blocking state
kernel: DEV: port 6(vb-xxx) entered disabled state
kernel: device vb-xxx entered promiscuous mode
systemd-machined[765]: New machine xxx.
```

The stop/start is the only thing that works consistently.

**Comment by clhedrick** [comment]
I’ve seen the same issue with lxc. It looks like a problem with Macvlan not releasing the Mac and iP address when the container exits. I’m moving to kvm because of this

**Comment by falstaff1288** [comment]
A working fix to counter the issue is to assign a random MAC address in the .network file that configures the mv-* interface. Example:

'''
[Match]
Name=mv*

[Link]
MACAddress=00:16:3e:03:94:24

[Network]
DHCP=ipv4
'''

**Comment by Andrei-Pozolotin** [comment]
the issue is still present as of `systemd 242 (242.84-2-arch)`

**Comment by amessina** [comment]
Correction to my comment from April 6 above:

This issue still occurs with the following version in both the host and container.
systemd 245 (v245.4-1.fc32)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified

However, some additional detail--I can only reproduce the issue when I run services inside the container that connect (FreeIPA, etc.).

If I just start the container without running any services, I can poweroff and restart the container without this issue.

**Comment by riaqn** [comment]
I confirm this issue still exists.

**Comment by srd424** [comment]
Seeing this with a focal (systemd 245) container on bionic..

**Comment by srd424** [comment]
There also seems to be some timing element here .. I restarted a container earlier, and it failed to bring the interface up. Several hours later, I have restarted it again and it succeeded. It's not something really weird like lingering sockets causing the problem, is it...?

**Comment by kakra** [comment]
That would explain why I only sometimes rarely still see it while most of the times it seems to be fixed. Also, only containers connecting macvlan are affected (tho, I'm using only either macvlan or no private network). All my interfaces are managed by systemd-networkd if that matters.

**Comment by srd424** [comment]
Just for clarification, the error returned by the kernel changed at some point in the last couple of years as well: https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg1883979.html

This is still causing me real problems, BTW!

**Comment by srd424** [comment]
OK, horrible bodge time. It seems to work, but I can't say why!

```
[Unit]
DefaultDependencies=no
After=final.target

[Service]
Type=oneshot
ExecStart=/usr/bin/perl -e 'open(my $fh, "ip -br link show type macvlan|"); while(<$fh>) { s/@.*$//; system("ip link del $_") }; '

[Install]
WantedBy=final.target
```
(Deletes all macvlan interfaces on container shutdown.)

 Any chance systemd-nspawn (or systemd in the container) could grow this functionality natively?

**Comment by srd424** [comment]
Hmm, spoke too soon, the bodge seems to improve things in some circumstances but isn't perfect :(

**Comment by keszybz** [comment]
The way that macvlans and other netdevs-that-require-other-netdevs are brought up has been significantly reworked by @yuwata. It would be great if you could check that this still occurs with v250+.

**Comment by srd424** [comment]
I will add that to the to-do list, but I suspect it's more to do with container shutdown. I've actually just stumbled across https://github.com/systemd/systemd/pull/3111, which sounds very similar to what I think needs to be done for MACVLANs..

**Comment by amessina** [comment]
I no longer see this issue using systemd-249.9-1.fc35.x86_64 which AFAICT is built from the v249-stable branch: https://github.com/systemd/systemd-stable/tree/v249-stable.

**Comment by mpfau** [comment]
still occurs with systemd-250.3-2~bpo11+1 (debian bullseye/backports)

**Comment by amessina** [comment]
Unfortunately I see this issue again using systemd-249.11-2.fc35.x86_64.  I will be upgrading to Fedora 36 shortly and will report again.

**Comment by mpfau** [comment]
@amessina did you have any success with Fedora 36?

**Comment by amessina** [comment]
@mpfau sorry for not reporting back initially.  I haven't had exhaustive testing, but in general use, I haven't seen this yet in Fedora 36.

**Comment by amessina** [comment]
@mpfau I just upgraded my machines to Fedora 37 and I can report this issue still exists with my systemd-nspawn instances in Fedora 37 systemd-251.8-586.fc37.x86_64.

**Comment by yuwata** [comment]
Note, the error code is changed from EBUSY to EADDRINUSE by https://github.com/torvalds/linux/commit/59f997b088d26a774958cb7b17b0763cd82de7ec.

**Comment by amessina** [comment]
This issue still occurs in the new update to Fedora 40 with systemd-255.4-1.fc40.x86_64.

**Comment by mpfau** [comment]
The issue also still occurs on debian bookworm with systemd 252 and kernel version 6.1.0-21-amd64.

**Comment by mpfau** [comment]
During the last couple weeks, it only occured on containers with multiple macvlan interfaces. We could reboot containers with a single macvlan interface without problems.

We ran into problems with containers with multiple macvlan interfaces (on the same physical interface). In those cases, the interfaces where down after the reboot of the container and we found msgs like `mv-ext: Could not bring up interface, ignoring: Address already in use` in systemd-networkd journal. 

At that point in time we always make sure that no process (that actually binds to an ip on that device) is active anymore. But even if there is none, invoking `ip link set mv-ext up` errors out in most cases. Sometimes another reboot of the container helps, sometimes not. In some cases, we had to reboot the bare metal to be able to bring the container up again.

Has anyone found a workaround for this issue? It seems like manually changing the MACs of those devices from within network units `[Link]` helps for one reboot, but on the next, the manually assigned MACs are not freed again and nspawn falls back to the auto-assigned MACs.

**Comment by kakra** [comment]
> Has anyone found a workaround for this issue?

I made exactly the same observations. Rarely, the issue still occurs with single-interface containers. My work-around is usually to reboot the container again, if it doesn't work: stop it, start it. If it still doesn't work: stop, wait 1-2 minutes, start. That last step usually works as a last resort, we didn't have to reboot the bare metal in a long time.

**Comment by mpfau** [comment]
It seems like nspawn does not delete the macvlan interfaces after the container shutdown and the kernel needs some time to actually do the cleanup. When the container is started again before the kernel actually cleaned up the orphan MACVLAN interfaces, it will fail to create another MACVLAN interface with the same mac address...

So we changed our containers to delete those interfaces on shutdown (e.g. `ip link del dev mv-ext`). This solved the issue.

Is there any reason why nspawn does not delete all network interfaces after shutdown?

**Comment by yuwata** [comment]
The suggestion by @mpfau is implemented by #33869, though still I cannot reproduce the issue on my laptop.

Could anyone test the patch?
